### PR TITLE
issue #5440 terraform crash in resource elasticsearch_domain

### LINKED ISSUE #5440: Terraform crash (SIGSEGV) when running import of elasticsearch with Error logs enabled
_This issue was originally opened by @jverhoeks as hashicorp/terraform#18595. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version

```
2018/08/03 11:01:21 [INFO] Terraform version: 0.11.7  41e50bd32a8825a84535e353c3674af8ce799161
2018/08/03 11:01:21 [INFO] Go runtime version: go1.10.1
```

### Terraform Configuration Files

```hcl
...
```

### Debug Output

[crash.gpg.log](https://github.com/hashicorp/terraform/files/2256800/crash.gpg.log)

### Crash Output

```
* import aws_elasticsearch_domain.services_client result: client: aws_elasticsearch_domain.services_client: unexpected EOF


panic: runtime error: invalid memory address or nil pointer dereference
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2b22db9]
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4:
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: goroutine 75 [running]:
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsElasticSearchDomainRead(0xc420323b20, 0x3616680, 0xc420076c00, 0xc420323b20, 0x0)
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_elasticsearch_domain.go:551 +0x1719
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Refresh(0xc42038eaf0, 0xc42060fb30, 0x3616680, 0xc420076c00, 0xc420534d38, 0xc4208dd201, 0x80000000018)
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:352 +0x17e
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Refresh(0xc4202b1d50, 0xc42060fae0, 0xc42060fb30, 0x6c186c8, 0x0, 0x18)
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:308 +0x9a
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Refresh(0xc420390b00, 0xc4208b4e70, 0xc4208b5020, 0x0, 0x0)
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:549 +0x4e
2018-08-03T10:59:01.118+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: reflect.Value.call(0xc420388c00, 0xc4204ce710, 0x13, 0x3b31969, 0x4, 0xc420044f20, 0x3, 0x3, 0x8, 0xc42002de80, ...)
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/usr/local/go/src/reflect/value.go:434 +0x905
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: reflect.Value.Call(0xc420388c00, 0xc4204ce710, 0x13, 0xc42002df20, 0x3, 0x3, 0xc42002df20, 0xc42002df28, 0xc4201f9840)
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/usr/local/go/src/reflect/value.go:302 +0xa4
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: net/rpc.(*service).call(0xc42037a800, 0xc420186190, 0xc42037c8d0, 0xc42018e700, 0xc420391680, 0x31495c0, 0xc4208b4e70, 0x16, 0x3149600, 0xc4208b5020, ...)
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/usr/local/go/src/net/rpc/server.go:381 +0x142
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: created by net/rpc.(*Server).ServeCodec
2018-08-03T10:59:01.119+0200 [DEBUG] plugin.terraform-provider-aws_v1.30.0_x4: 	/usr/local/go/src/net/rpc/server.go:475 +0x36b
2018/08/03 10:59:01 [ERROR] root: eval: *terraform.EvalRefresh, err: aws_elasticsearch_domain.services_client: unexpected EOF
2018/08/03 10:59:01 [ERROR] root: eval: *terraform.EvalSequence, err: aws_elasticsearch_domain.services_client: unexpected EOF
```

### Expected Behavior

Resource imported

### Actual Behavior

Nothing

### Steps to Reproduce

1. terraform import  aws_elasticsearch_domain.services_client client

### Additional Context

It worked before with elasticsearch instances. Only difference is this one has a bigger disksize (200gb vs 100gb before)

### PR DESCRIPTION
**Pull request by saravanan30erd**
Fixes #5440 

Changes proposed in this pull request:

* Change 1
fixed panic nil pointer dereference error.
* Change 2
Properly dereferenced other attributes too.

### CODE REVIEW

File: aws/resource_aws_elasticsearch_domain.go [hunk 1]
```diff
@@ -548,7 +548,7 @@ func resourceAwsElasticSearchDomainRead(d *schema.ResourceData, meta interface{}
 	}
 	if ds.SnapshotOptions != nil {
```
Comment (bflad) [inline aws/resource_aws_elasticsearch_domain.go]: This `d.Set()` was incorrectly implemented before (notice the missing `[]` before the map) and we should probably fix it now. I think the AutomatedSnapshotStartHour value also needs to be wrapped in `int()`. Something like the below could handle both for now and return an error if incorrect:

```go
if err := d.Set("snapshot_options", flattenESSnapshotOptions(ds.SnapshotOptions)); err != nil {
	return fmt.Errorf("error setting snapshot_options": err)
}
```

```go
func flattenESSnapshotOptions(snapshotOptions *elasticsearch.SnapshotOptions) []map[string]interface{} {
	if snapshotOptions == nil {
		return []map[string]interface{}{}
	}

	m := map[string]interface{}{
		"automated_snapshot_start_hour": int(aws.Int64Value(snapshotOptions.AutomatedSnapshotStartHour)),
	}

	return []map[string]interface{}{m}
}
```

**Comment by saravanan30erd** [comment]
@bflad Done 👍

**Comment by bflad** [comment]
Looks like one more little thing if you have a chance, since we actually fixed the "container" `snapshot_options` attribute, its reporting a difference if its not defined in the configuration:

```
=== RUN   TestAccAWSElasticSearchDomain_basic
--- FAIL: TestAccAWSElasticSearchDomain_basic (1214.92s)
	testing.go:518: Step 0 error: After applying this step, the plan was not empty:
		
		DIFF:
		
		UPDATE: aws_elasticsearch_domain.example
		  snapshot_options.#: "1" => "0"
```

It should be okay if we tell the schema to ignore this difference as the underlying attributes are what really matter in this case. 👍 

```go
// In "snapshot_options" attribute definition:
DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
	if old == "1" && new == "0" {
		return true
	}
	return false
},
```

Thanks so much!

**Comment by saravanan30erd** [comment]
@bflad I added diffsuppress func. I think if we check the nil value for `SnapshotOptions` before setting it like below, we can prevent this issue right?

```
if ds.SnapshotOptions != nil {
    if err := d.Set("snapshot_options", flattenESSnapshotOptions(ds.SnapshotOptions)); err != nil 
         {
	        return fmt.Errorf("error setting snapshot_options: %s", err)
	 }
}
```

**Pull request merged**